central/reports/service/v2/service_impl.go

```diff
@@ -191,3 +196,60 @@ func (s *serviceImpl) CancelReport(ctx context.Context, req *apiV2.ResourceByID)
 
 	return &apiV2.Empty{}, nil
 }
+
+func (s *serviceImpl) DownloadReport(ctx context.Context, req *apiV2.DownloadReportRequest) (*apiV2.DownloadReportResponse, error) {
+	if req == nil || req.GetReportJobId() == "" {
+		return nil, errors.Wrap(errox.InvalidArgs, "Empty request or report job id")
+	}
+
+	slimUser := authn.UserFromContext(ctx)
+	if slimUser == nil {
+		return nil, errors.New("Could not determine user identity from provided context")
+	}
+
+	rep, found, err := s.snapshotDatastore.Get(ctx, req.GetReportJobId())
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error finding report snapshot with job ID '%s'.", req.GetReportJobId())
+	}
+
+	if !found {
+		return nil, errors.Wrapf(errox.NotFound, "Error finding report snapshot with job ID '%s'.", req.GetReportJobId())
+	}
+
+	if slimUser.GetId() != rep.GetRequester().GetId() {
+		return nil, errors.Wrap(errox.NotAuthorized, "Report cannot be downloaded by a user who did not request the report.")
+	}
+
+	status := rep.GetReportStatus()
+	if status.GetReportNotificationMethod() != storage.ReportStatus_DOWNLOAD {
+		return nil, errors.Wrapf(errox.InvalidArgs, "Report download is not requested for job id %q", req.GetReportJobId())
```

Comment:
This is somewhat unclear. Maybe `Report job id %q did not generate a downloadable report and hence report cannot be downloaded.`